File: ios/NitroGoogleCast/Converters/GCKMediaMetadata+toMediaMetadata.swift
import Foundation
import GoogleCast
import NitroModules

/// Converts a Google Cast `GCKMediaMetadata` into a generated `MediaMetadata` struct.
///
/// Reverse of `MediaMetadata+toGckMediaMetadata.swift`. Each standard field is read with the
/// GCK field's required typed getter (string / integer / double / date) guarded by
/// `containsKey:`, so a field that is genuinely absent stays `nil` (integer getters would
/// otherwise return `0` for missing fields).
///
/// Standard-vs-custom key rule: a key is "standard" iff it is one of the predefined
/// `kGCKMetadataKey*` constants that maps to a struct field (`Self.standardKeys`). Every
/// other key present in `allKeys` is application-defined and is collected into `customData`.
/// This partition is symmetric with the forward converter and never double-counts a key.
/// Empty collections normalize to `nil` (GCK reports an empty images array / no custom keys
/// for absent data), keeping round-trips stable.
extension GCKMediaMetadata {
  func toMediaMetadata() -> MediaMetadata {
    let mappedImages = (images() as? [GCKImage])?.map { $0.toWebImage() } ?? []

    return MediaMetadata(
      type: metadataType.toMediaMetadataType(),
      images: mappedImages.isEmpty ? nil : mappedImages,
      title: stringIfPresent(kGCKMetadataKeyTitle),
      subtitle: stringIfPresent(kGCKMetadataKeySubtitle),
      artist: stringIfPresent(kGCKMetadataKeyArtist),
      releaseDate: dateStringIfPresent(kGCKMetadataKeyReleaseDate),
      studio: stringIfPresent(kGCKMetadataKeyStudio),
      albumTitle: stringIfPresent(kGCKMetadataKeyAlbumTitle),
      albumArtist: stringIfPresent(kGCKMetadataKeyAlbumArtist),
      composer: stringIfPresent(kGCKMetadataKeyComposer),
      discNumber: integerIfPresent(kGCKMetadataKeyDiscNumber),
      trackNumber: integerIfPresent(kGCKMetadataKeyTrackNumber),
      creationDate: dateStringIfPresent(kGCKMetadataKeyCreationDate),
      location: stringIfPresent(kGCKMetadataKeyLocationName),
      latitude: doubleIfPresent(kGCKMetadataKeyLocationLatitude),
      longitude: doubleIfPresent(kGCKMetadataKeyLocationLongitude),
      width: integerIfPresent(kGCKMetadataKeyWidth),
      height: integerIfPresent(kGCKMetadataKeyHeight),
      broadcastDate: dateStringIfPresent(kGCKMetadataKeyBroadcastDate),
      episodeNumber: integerIfPresent(kGCKMetadataKeyEpisodeNumber),
      seasonNumber: integerIfPresent(kGCKMetadataKeySeasonNumber),
      seriesTitle: stringIfPresent(kGCKMetadataKeySeriesTitle),
      customData: extractCustomData()
    )
  }

  // MARK: - Typed presence-guarded readers

  private func stringIfPresent(_ key: String) -> String? {
    return containsKey(key) ? string(forKey: key) : nil
  }

  private func dateStringIfPresent(_ key: String) -> String? {
    return containsKey(key) ? dateAsString(forKey: key) : nil
  }

  private func integerIfPresent(_ key: String) -> Double? {
    return containsKey(key) ? Double(integer(forKey: key)) : nil
  }

  private func doubleIfPresent(_ key: String) -> Double? {
    return containsKey(key) ? double(forKey: key) : nil
  }

  // MARK: - Custom data partition

  /// The predefined keys consumed by struct fields. Anything else is application-defined.
  private static let standardKeys: Set<String> = [
    kGCKMetadataKeyTitle, kGCKMetadataKeySubtitle, kGCKMetadataKeyArtist,
    kGCKMetadataKeyReleaseDate, kGCKMetadataKeyStudio, kGCKMetadataKeyAlbumTitle,
    kGCKMetadataKeyAlbumArtist, kGCKMetadataKeyComposer, kGCKMetadataKeyDiscNumber,
    kGCKMetadataKeyTrackNumber, kGCKMetadataKeyCreationDate, kGCKMetadataKeyLocationName,
    kGCKMetadataKeyLocationLatitude, kGCKMetadataKeyLocationLongitude, kGCKMetadataKeyWidth,
    kGCKMetadataKeyHeight, kGCKMetadataKeyBroadcastDate, kGCKMetadataKeyEpisodeNumber,
    kGCKMetadataKeySeasonNumber, kGCKMetadataKeySeriesTitle,
  ]

  private func extractCustomData() -> AnyMap? {
    let customKeys = allKeys().filter { !Self.standardKeys.contains($0) }
    if customKeys.isEmpty {
      return nil
    }
    let map = AnyMap()
    for key in customKeys {
      let value = object(forKey: key)
      if let string = value as? String {
        map.setString(key: key, value: string)
      } else if let number = value as? NSNumber {
        map.setDouble(key: key, value: number.doubleValue)
      }
    }
    return map
  }
}

File: android/src/main/java/com/margelo/nitro/googlecast/converters/GckMediaMetadata+toMediaMetadata.kt
package com.margelo.nitro.googlecast.converters

import com.margelo.nitro.core.AnyMap
import com.margelo.nitro.googlecast.MediaMetadata
import com.margelo.nitro.googlecast.MediaMetadataType
import com.margelo.nitro.googlecast.WebImage
import com.google.android.gms.cast.MediaMetadata as GckMediaMetadata

/**
 * Converts a Google Cast [GckMediaMetadata] into a generated [MediaMetadata] struct.
 *
 * Reverse of `MediaMetadata+toGckMediaMetadata.kt`. Each standard field is read with the GCK
 * field's typed getter guarded by `containsKey` so a genuinely absent field stays `null`
 * (int/double getters otherwise default to `0`).
 *
 * Standard-vs-custom key rule: a key is "standard" iff it is one of the predefined `KEY_*`
 * constants that maps to a struct field (see [STANDARD_KEYS]). Every other key in `keySet()`
 * is application-defined and is collected into `customData`. GCK Android exposes no generic
 * value getter for arbitrary keys, so custom values are read best-effort as string then
 * double — full fidelity is deferred to the instrumented parity pass.
 */
internal fun GckMediaMetadata.toMediaMetadata(): MediaMetadata {
  val mappedImages = images?.map { it.toWebImage() } ?: emptyList<WebImage>()

  return MediaMetadata(
    type = mediaMetadataTypeFromGck(mediaType),
    images = if (mappedImages.isEmpty()) null else mappedImages.toTypedArray(),
    title = stringIfPresent(GckMediaMetadata.KEY_TITLE),
    subtitle = stringIfPresent(GckMediaMetadata.KEY_SUBTITLE),
    artist = stringIfPresent(GckMediaMetadata.KEY_ARTIST),
    releaseDate = dateStringIfPresent(GckMediaMetadata.KEY_RELEASE_DATE),
    studio = stringIfPresent(GckMediaMetadata.KEY_STUDIO),
    albumTitle = stringIfPresent(GckMediaMetadata.KEY_ALBUM_TITLE),
    albumArtist = stringIfPresent(GckMediaMetadata.KEY_ALBUM_ARTIST),
    composer = stringIfPresent(GckMediaMetadata.KEY_COMPOSER),
    discNumber = intIfPresent(GckMediaMetadata.KEY_DISC_NUMBER),
    trackNumber = intIfPresent(GckMediaMetadata.KEY_TRACK_NUMBER),
    creationDate = dateStringIfPresent(GckMediaMetadata.KEY_CREATION_DATE),
    location = stringIfPresent(GckMediaMetadata.KEY_LOCATION_NAME),
    latitude = doubleIfPresent(GckMediaMetadata.KEY_LOCATION_LATITUDE),
    longitude = doubleIfPresent(GckMediaMetadata.KEY_LOCATION_LONGITUDE),
    width = intIfPresent(GckMediaMetadata.KEY_WIDTH),
    height = intIfPresent(GckMediaMetadata.KEY_HEIGHT),
    broadcastDate = dateStringIfPresent(GckMediaMetadata.KEY_BROADCAST_DATE),
    episodeNumber = intIfPresent(GckMediaMetadata.KEY_EPISODE_NUMBER),
    seasonNumber = intIfPresent(GckMediaMetadata.KEY_SEASON_NUMBER),
    seriesTitle = stringIfPresent(GckMediaMetadata.KEY_SERIES_TITLE),
    customData = extractCustomData()
  )
}

private fun GckMediaMetadata.stringIfPresent(key: String): String? =
  if (containsKey(key)) getString(key) else null

private fun GckMediaMetadata.dateStringIfPresent(key: String): String? =
  if (containsKey(key)) getDateAsString(key) else null

private fun GckMediaMetadata.intIfPresent(key: String): Double? =
  if (containsKey(key)) getInt(key).toDouble() else null

private fun GckMediaMetadata.doubleIfPresent(key: String): Double? =
  if (containsKey(key)) getDouble(key) else null

private val STANDARD_KEYS: Set<String> = setOf(
  GckMediaMetadata.KEY_TITLE, GckMediaMetadata.KEY_SUBTITLE, GckMediaMetadata.KEY_ARTIST,
  GckMediaMetadata.KEY_RELEASE_DATE, GckMediaMetadata.KEY_STUDIO, GckMediaMetadata.KEY_ALBUM_TITLE,
  GckMediaMetadata.KEY_ALBUM_ARTIST, GckMediaMetadata.KEY_COMPOSER, GckMediaMetadata.KEY_DISC_NUMBER,
  GckMediaMetadata.KEY_TRACK_NUMBER, GckMediaMetadata.KEY_CREATION_DATE, GckMediaMetadata.KEY_LOCATION_NAME,
  GckMediaMetadata.KEY_LOCATION_LATITUDE, GckMediaMetadata.KEY_LOCATION_LONGITUDE, GckMediaMetadata.KEY_WIDTH,
  GckMediaMetadata.KEY_HEIGHT, GckMediaMetadata.KEY_BROADCAST_DATE, GckMediaMetadata.KEY_EPISODE_NUMBER,
  GckMediaMetadata.KEY_SEASON_NUMBER, GckMediaMetadata.KEY_SERIES_TITLE
)

private fun GckMediaMetadata.extractCustomData(): AnyMap? {
  val customKeys = keySet().filter { it !in STANDARD_KEYS }
  if (customKeys.isEmpty()) {
    return null
  }
  val map = AnyMap()
  for (key in customKeys) {
    val asString = getString(key)
    if (asString != null) {
      map.setString(key, asString)
    } else {
      map.setDouble(key, getDouble(key))
    }
  }
  return map
}

private fun mediaMetadataTypeFromGck(value: Int): MediaMetadataType =
  when (value) {
    GckMediaMetadata.MEDIA_TYPE_GENERIC -> MediaMetadataType.GENERIC
    GckMediaMetadata.MEDIA_TYPE_MOVIE -> MediaMetadataType.MOVIE
    GckMediaMetadata.MEDIA_TYPE_MUSIC_TRACK -> MediaMetadataType.MUSICTRACK
    GckMediaMetadata.MEDIA_TYPE_PHOTO -> MediaMetadataType.PHOTO
    GckMediaMetadata.MEDIA_TYPE_TV_SHOW -> MediaMetadataType.TVSHOW
    else -> if (value >= GckMediaMetadata.MEDIA_TYPE_USER) MediaMetadataType.USER else MediaMetadataType.GENERIC
  }
